controllers/front/listing/CategoryController.php

```diff
@@ -45,6 +45,13 @@ public function canonicalRedirection($canonicalURL = '')
     {
         if (Validate::isLoadedObject($this->category)) {
             parent::canonicalRedirection($this->context->link->getCategoryLink($this->category));
+        } elseif (!Validate::isLoadedObject($this->category) && $canonicalURL == '') {
+            header('HTTP/1.1 404 Not Found');
+            header('Status: 404 Not Found');
+            $this->errors[] = $this->trans("This category doesn't exists.", [], 'Shop.Notifications.Error');
+            $this->setTemplate('errors/404');
+
+            return;
         } elseif ($canonicalURL) {
             parent::canonicalRedirection($canonicalURL);
         }
```

Comment:
```suggestion
        } elseif ($canonicalURL) {
            parent::canonicalRedirection($canonicalURL);
        } else {
            header('HTTP/1.1 404 Not Found');
            header('Status: 404 Not Found');
            $this->errors[] = $this->trans("This category doesn't exists.", [], 'Shop.Notifications.Error');
            $this->setTemplate('errors/404');
        } 
```

Because if the object is valid, it goes to the first condition. If there is a canonical URL it goes to the second, else it must be a 404 because it's not loaded and there is no canonical URL